### PR TITLE
re-enable wallet in autoprune

### PR DESCRIPTION
**Pull request by jonasschnelli**

### CODE REVIEW

**Comment by morcos** [comment]
Nice! I'm glad you looked at this.

The way you checked `pindexRescan` isn't sufficient however.  You can't assume a pruned node has all the blocks past a certain point.  Blocks are deleted from disk in the order they were written, which doesn't necessarily correspond with order in the chain.  But perhaps you could scan backwards from the Tip until you get to either `pindexRescan` or one which isn't BLOCK_HAVE_DATA.

**Comment by jonasschnelli** [comment]
Right. I did ignore the fact that blocks are not fly in in order. Will change it to your proposed way of scanning backwards.

Same needs to be done for #6058

File: src/init.cpp [hunk 1]
```diff
@@ -703,16 +703,12 @@ bool AppInit2(boost::thread_group& threadGroup)
...
         if (GetBoolArg("-txindex", false))
             return InitError(_("Prune mode is incompatible with -txindex."));
 #ifdef ENABLE_WALLET
-        if (!GetBoolArg("-disablewallet", false)) {
-            if (SoftSetBoolArg("-disablewallet", true))
-                LogPrintf("%s : parameter interaction: -prune -> setting -disablewallet=1\n", __func__);
-            else
-                return InitError(_("Can't run with a wallet in prune mode."));
+        if (GetBoolArg("-rescan", false)) {
+            return InitError(_("Rescans are not possible in pruned mode. You might use -reindex which then download the whole blockchain again."));
```
Comment (PRabahy) [inline src/init.cpp]: This message is unclear to me. Do you mean:
"Rescans are not possible in pruned mode. You will need to use -reindex which will download the whole blockchain again."?

Comment (jonasschnelli) [inline src/init.cpp]: Thanks!
English is not my mother language and I always struggle with it.
Feel free to also correct/rewrite the other warnings and errors (as well as the code comments).

**Comment by jonasschnelli** [comment]
Fixed `pindexRescan` check mentioned by @morcos.

**Comment by sipa** [comment]
Concept & code review ACK, but this definitely needs testing.

**Comment by jonasschnelli** [comment]
Also applied an additional test during wallet catchup-rescan if the previous block has TXs but no data to be sure we have pruned this block.

**Comment by jonasschnelli** [comment]
Rebased.

File: src/init.cpp [hunk 2]
```diff
@@ -714,16 +714,12 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
         if (GetBoolArg("-txindex", false))
             return InitError(_("Prune mode is incompatible with -txindex."));
 #ifdef ENABLE_WALLET
-        if (!GetBoolArg("-disablewallet", false)) {
-            if (SoftSetBoolArg("-disablewallet", true))
-                LogPrintf("%s : parameter interaction: -prune -> setting -disablewallet=1\n", __func__);
-            else
-                return InitError(_("Can't run with a wallet in prune mode."));
+        if (GetBoolArg("-rescan", false)) {
+            return InitError(_("Rescans are not possible in pruned mode. You will need to use -reindex which will download the whole blockchain again."));
```
Comment (luke-jr) [inline src/init.cpp]: What prevents this error if the user has passed -reindex? In particular, a pruned node may not have the space for the whole blockchain, and may need to redownload it with pruning enabled.

(... btw, does -reindex actually work correctly with pruned data? I think last time I tried, it ignored blocks after the first missing one, and redownloaded/reappended them...)

**Comment by laanwj** [comment]
utACK.

**Pull request merged**